---
title: "Pangenome analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load(data.table,
               tidyverse,
               ggplot2,
               PFAM.db,
               ComplexHeatmap)
```

# Load
```{r, results=F, message=F}
redcap.DM <- readxl::read_xlsx("../High_Prev_Metadata_redcap.xlsx",sheet=1) %>% as.data.frame()
redcap.LG <- readxl::read_xlsx("../High_Prev_Metadata_redcap.xlsx",sheet=2) %>% as.data.frame()

metadata <- fread("../metadata.txt") %>% as.data.frame()
rownames(metadata) <- metadata$PID

genes <- read.csv("../hiv.t32.p_copri_panphlan_prof/gene_presence_absence_6.tsv", 
               sep="\t"
               ) %>%
  as.data.frame()

```

## Add sexual behavior to metadata
```{r}
sex.behav.q <- c("In the past 6 months, what is the frequency at which you have had receptive anal intercourse?",
                 "What gender do you identify with?",
                 "What is your sexual orientation?"
                 )

metadata <- merge(
    metadata,
    rbind(redcap.DM[,c("Record ID", sex.behav.q)],
          redcap.LG[,c("Record ID", sex.behav.q)]),
    by.x="PID",
    by.y="Record ID")
rownames(metadata) <- metadata$PID

metadata$RAI_freq <- metadata$`In the past 6 months, what is the frequency at which you have had receptive anal intercourse?`
metadata$`In the past 6 months, what is the frequency at which you have had receptive anal intercourse?` <- NULL
metadata$RAI <- metadata$RAI_freq!="Never"

metadata$Gender <- metadata$`What gender do you identify with?`
metadata$Orientation <- metadata$`What is your sexual orientation?`
metadata$MSM <- metadata$Gender=="Men" & (metadata$Orientation=="Gay" | metadata$Orientation=="Bisexual")

# address missing data with follow-up from Mallory
metadata[metadata$PID=="DM001", "MSM"] <- TRUE
metadata[metadata$PID=="DM002", "MSM"] <- TRUE
metadata[metadata$PID=="DM007", "MSM"] <- TRUE

metadata[metadata$PID=="DM093", "MSM"] <- FALSE
```

# Heatmap {.tabset}

```{r}
#rownames(genes) <- genes$X

# Make this a matrix for the heatmap
genes.mat <- genes %>% 
    dplyr::select(-c(X, annotation)) %>% 
    as.matrix()
# Remove the rows with NA values - these rows are the unannotated rows
genes.mat <- genes.mat[which(rowSums(is.na(genes.mat)) == 0),]
if(sum(is.na(genes.mat)) != 0){
    print("This matrix has NA values!")
}
colnames(genes.mat) <- colnames(genes.mat) %>% 
    gsub(pattern="_p_copri.csv", replacement="")
```

## Risk status
```{r}
group <- metadata[colnames(genes.mat), "Risk_Status"]
group.colors <- group
group.colors[group=="Low"] <- "blue"
group.colors[group=="High"] <- "yellow"
group.colors[group=="Positive"] <- "red"
group.colors[is.na(group)] <- "black"

# Based on risk status
heatmap(genes.mat, labRow=F, na.rm=T,
        ColSideColors=group.colors,
        scale="none", col=pals::coolwarm(n = 2))

```

## RAI
```{r}
metadata[is.na(metadata$RAI), "RAI"] <- "No.Info"
RAI <- metadata[colnames(genes.mat), "RAI"]
RAI.colors <- RAI
RAI.colors[RAI=="TRUE"] <- "purple"
RAI.colors[RAI=="FALSE"] <- "yellow"
RAI.colors[RAI=="No.Info"] <- "white"
RAI.colors[is.na(RAI)] <- "black"

# Based on risk status
p <- heatmap(genes.mat, labRow=F, na.rm=T,
        ColSideColors=RAI.colors,
        scale="none", col=pals::coolwarm(n = 2))

```

## RAI frequency
```{r}
metadata[is.na(metadata$RAI_freq), "RAI_freq"] <- "No.Info"
RAI.freq <- metadata[colnames(genes.mat), "RAI_freq"]
RAI.freq.colors <- RAI.freq
RAI.freq.colors[RAI.freq==">1 time per week"] <- "purple"
RAI.freq.colors[RAI.freq=="< 1 time per week"] <- "gray"
RAI.freq.colors[RAI.freq=="Never"] <- "yellow"
RAI.freq.colors[RAI.freq=="No.Info"] <- "white"
RAI.freq.colors[is.na(RAI.freq)] <- "black"

# Based on risk status
heatmap(genes.mat, labRow=F, na.rm=T,
        ColSideColors=RAI.freq.colors,
        scale="none", col=pals::coolwarm(n = 2))


```

## MSM status
```{r}
metadata[is.na(metadata$MSM), "MSM"] <- "No.Info"
MSM <- metadata[colnames(genes.mat), "MSM"]

MSM.colors <- MSM
MSM.colors[MSM=="TRUE"] <- "purple"
MSM.colors[MSM=="FALSE"] <- "yellow"
MSM.colors[MSM=="No.Info"] <- "white"
MSM.colors[is.na(MSM)] <- "black"

MSM[is.na(MSM)] <- "Reference"
MSM[MSM=="TRUE"] <- "MSM"
MSM[MSM=="FALSE"] <- "MSW"

# Based on risk status
#pdf("../figures/p_copri_pangenome_heatmap_MSM.pdf")
heatmap(genes.mat, labRow=F, na.rm=T,
        ColSideColors=MSM.colors,
        scale="none", col=pals::coolwarm(n = 2)) 
#dev.off()

genes.mat.heatmap <- genes.mat
samp.names <- colnames(genes.mat)
samp.names[str_starts(samp.names, "REF")] <- paste("Strain", 1:4)
colnames(genes.mat.heatmap) <- samp.names

annotation_col <- as.data.frame(MSM, row.names=colnames(genes.mat.heatmap))
annotation_col[annotation_col$MSM=="No.Info", "MSM"] <- NA
annotation_colors <- c("MSW"="yellow", "MSM"="purple", "Reference"="black")


pheatmap::pheatmap(genes.mat.heatmap, annotation_col=annotation_col, 
                   annotation_colors=list(MSM=annotation_colors),
                   color=c("blue", "red"), legend_breaks=c(0,1), legend_labels=c("Absent", "Present"),
                   alpha=1,
                   file="../figures/p_copri_pangenome_heatmap_MSM.pdf",
                   width=10, height=10
                   )

```

## HIV status
```{r}
metadata[is.na(metadata$HIV_Status), "HIV_Status"] <- "No.Info"
HIV <- metadata[colnames(genes.mat), "HIV_Status"]

HIV.colors <- HIV
HIV.colors[HIV=="Positive"] <- "purple"
HIV.colors[HIV=="Negative"] <- "yellow"
HIV.colors[HIV=="No.Info"] <- "white"
HIV.colors[is.na(HIV)] <- "black"

# Based on risk status
heatmap(genes.mat, labRow=F, na.rm=T,
        ColSideColors=HIV.colors,
        scale="none", col=pals::coolwarm(n = 2))

```

# Differential abundance (PFams) {.tabset}

## PFAMS {.tabset}
```{r}
# remove rows with NA - these are unannotated or messed up from unannotated values
genes.nona <- genes[which(rowSums(is.na(genes)) == 0),]

DA.gene.data <- 
    genes.nona %>% 
    group_by(annotation) %>%
    summarise_at(vars(-c(group_cols(), X)), 
                 mean)
colnames(DA.gene.data) <- colnames(DA.gene.data) %>% 
    gsub(pattern="_p_copri.csv", replacement="")


# heatmap using pfams
heatmap(DA.gene.data %>% dplyr::select(-annotation) %>% as.matrix(), 
        labRow=F, na.rm=T,
        ColSideColors=RAI.colors,
        scale="none", col=pals::coolwarm(n = 2))


run_t_tests <- function(df, 
                        metadata, metadata_column_name, 
                        condition_1, condition_2){
    return.df <- data.frame(matrix(nrow=nrow(df), ncol=3))
    colnames(return.df) <- c("Feature", "statistic", "p")
    
    for (i in 1:nrow(df)){
        return.df[i, "Feature"] <- df[i, "annotation"]
        
        curr.feature <- df[i,] %>% dplyr::select(-annotation)
        curr.feature <- t(curr.feature)
        sample.info <- metadata[rownames(curr.feature), metadata_column_name]
        
        t.test.res <- try(
            t.test(x=curr.feature[sample.info==condition_1],
                   y=curr.feature[sample.info==condition_2]),
            silent=TRUE
                        )
        if (is(t.test.res, "try-error")){
            return.df[i, "statistic"] <- 0
            return.df[i, "p"] <- 1
        } else {
            return.df[i, "statistic"] <- t.test.res$statistic
            return.df[i, "p"] <- t.test.res$p.value
        }
    }
    
    return.df <- return.df[order(return.df$p),]
    return.df$adjusted.p <- p.adjust(return.df$p, method="fdr")
    
    return(return.df)
}

map_pfs_to_domain <- function(x, PFDB){
    splitted <- strsplit(x, ",") %>% unlist()
    domains <- as.list(PFDB)[splitted] %>% as.character() %>% as.vector()
    return(paste(domains, collapse="; "))
}
```

### RAI
```{r}
RAI.DA <- run_t_tests(DA.gene.data %>% as.data.frame(), 
                      metadata, "RAI",
                      "TRUE", "FALSE")

RAI.DA <- RAI.DA[order(RAI.DA$p),]
RAI.DA$adjusted.p <- p.adjust(RAI.DA$p, method="BH")
    
RAI.DA[1:100, "Domain"] <- sapply(RAI.DA[1:100, "Feature"], 
                                                     function(x) map_pfs_to_domain(x, PFAMDE), USE.NAMES=F)


RAI.DA %>% head(100) %>% knitr::kable()
```

### MSM
```{r}
MSM.DA <- run_t_tests(DA.gene.data %>% as.data.frame(), 
                      metadata, "MSM",
                      "TRUE", "FALSE")

MSM.DA <- MSM.DA[order(MSM.DA$p),]
MSM.DA$adjusted.p <- p.adjust(MSM.DA$p, method="BH")
    
MSM.DA[1:100, "Domain"] <- sapply(MSM.DA[1:100, "Feature"], 
                                                     function(x) map_pfs_to_domain(x, PFAMDE), USE.NAMES=F)


MSM.DA %>% head(100) %>% knitr::kable()
```

### HIV
```{r}
HIV.DA <- run_t_tests(DA.gene.data %>% as.data.frame(), 
                      metadata, "HIV_Status",
                      "Positive", "Negative")

HIV.DA <- HIV.DA[order(HIV.DA$p),]
HIV.DA$adjusted.p <- p.adjust(HIV.DA$p, method="BH")
    
HIV.DA[1:100, "Domain"] <- sapply(HIV.DA[1:100, "Feature"], 
                                                     function(x) map_pfs_to_domain(x, PFAMDE), USE.NAMES=F)


HIV.DA %>% head(100) %>% knitr::kable()
```

# Reference Strains
```{r}
phylophlan.db <- fread("../phylophlan_databases/SGB.Jul20.txt.bz2") %>% as.data.frame()
REF.ids <- colnames(genes)[grep("REF",colnames(genes))]
REF.ids <- gsub(x=REF.ids, pattern="REF_", replacement="")

ref.1 <- phylophlan.db[grep(x=phylophlan.db$`List of reference genomes`,pattern=REF.ids[1]),]
print(REF.ids[1]); print(ref.1$`Assigned taxonomy`)

ref.2 <- phylophlan.db[grep(x=phylophlan.db$`List of reference genomes`,pattern=REF.ids[2]),]
print(REF.ids[2]); print(ref.2$`Assigned taxonomy`)

ref.3 <- phylophlan.db[grep(x=phylophlan.db$`List of reference genomes`,pattern=REF.ids[3]),]
print(REF.ids[3]); print(ref.3$`Assigned taxonomy`)

ref.4 <- phylophlan.db[grep(x=phylophlan.db$`List of reference genomes`,pattern=REF.ids[4]),]
print(REF.ids[4]); print(ref.4$`Assigned taxonomy`)
```