---
title: "Prevotella MAGs Phylogeny"
author: "John Sterrett"
date: "2023-04-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)


pacman::p_load(treeio,
               ggtree,
               ggplot2,
               tidyverse,
               data.table)
```

# Load
```{r}
metadata <- fread("../metadata.txt") %>% as.data.frame()
rownames(metadata) <- metadata$PID
tree <- read.tree("../Prevotella_phylogeny_04_17/RAxML_bestTree.phylophlan_input_bins_refined.tre")
bins.info <- fread("../Prevotella_phylogeny_04_17/hiv.t32.n40.metaspades.metabat2.checkm.drep.phylophlan.tsv") %>% 
    as.data.frame()


extract_genome_info <- function(bin.info, field){
    # will split this format -- [u|k]_[S|G|F]GBid:taxa_level:taxonomy:avg_dist
    splitted <- str_split(bin.info, pattern=":")
    splitted <- unlist(splitted)
    return (splitted[field])  
}

bins.info$PhyloPhlanID <- sapply(bins.info[,2], FUN=function(x) extract_genome_info(x, 1))
bins.info$Known <- sapply(bins.info$PhyloPhlanID, FUN=function(x) str_starts(x, "k"))
bins.info$ClassDepth <- sapply(bins.info[,2], FUN=function(x) extract_genome_info(x, 2))
bins.info$Taxonomy <- sapply(bins.info[,2], FUN=function(x) extract_genome_info(x, 3))
bins.info$AvgDist <- sapply(bins.info[,2], FUN=function(x) extract_genome_info(x, 4))
bins.info$tip <- bins.info$`#input_bin`
bins.info$found <- T
bins.info$PlotName <- sapply(bins.info$Taxonomy, FUN=function(x) str_split(x, pattern="\\|")[[1]][7])
bins.info[bins.info$Known==F, "PlotName"] <- NA


ref.bins.info <- phylophlan.db <- fread("../phylophlan_databases/SGB.Jul20.txt.bz2") %>% as.data.frame()
```

```{r}
tip.labels <- tree$tip.label
tree.bins.info <- bins.info[bins.info$tip %in% tip.labels,] %>%
    dplyr::select(c(tip, Taxonomy, PlotName))

extract_refgenome_info <- function(ref.id, ref.bins.info){
    # First, look for this genome as an SGB centroid
    genome.info <- ref.bins.info[grep(pattern=ref.id,
                                      x=ref.bins.info$`SGB Centroid`),]
    if (nrow(genome.info)>0){
        return(genome.info)
    }
    # If not found, look for it as a member of the reference genomes
    genome.info <- ref.bins.info[grep(pattern=ref.id,
                                      x=ref.bins.info$`List of reference genomes`),]
    if (nrow(genome.info)>0){
        return(genome.info)
    }
    # If not there, look for it in reconstructed genomes
    genome.info <- ref.bins.info[grep(pattern=ref.id,
                                      x=ref.bins.info$`List of reconstructed genomes`),]
    return(genome.info)
}

for(i in 1:length(tip.labels)){
    tip.name <- tip.labels[i]
    if(str_starts(tip.name, "GCA")){
        
        ref.info <- extract_refgenome_info(tip.name, ref.bins.info)
        tip.taxonomy <- ref.info$`Assigned taxonomy`
        tip.plotname <- str_split(tip.taxonomy, pattern="\\|")[[1]][7]
        new.row <- c(tip.name, tip.taxonomy, tip.plotname)
        names(new.row) <- c("tip", "Taxonomy", "PlotName")
        tree.bins.info <- rbind(tree.bins.info,
                                new.row)
    }
}

tree.bins.info[str_starts(tree.bins.info$tip, "GCA")==F, "Found"] <- TRUE
tree.bins.info[str_starts(tree.bins.info$tip, "GCA"), "Found"] <- NA

species.list <- c()
for (i in 1:nrow(tree.bins.info)){
    curr.species <- tree.bins.info[i, "PlotName"]
    # if it's the first occurrence of its species
    if (curr.species %in% species.list == F){
        species.list <- c(species.list, curr.species)
        tree.bins.info[i, "Subset.Name"] <- curr.species
    }
}

tree.bins.info[grepl(x=tree.bins.info$Subset.Name, pattern="SGB"), 
               "Subset.Name"] <- NA
tree.bins.info[grepl(x=tree.bins.info$Subset.Name, pattern="_sp"), 
               "Subset.Name"] <- NA
# Give each of our MAGs a number
tree.bins.info[grepl(tree.bins.info$tip, pattern="GCA")==F, "MAG.num"] <- 1:sum(grepl(tree.bins.info$tip, pattern="GCA")==F)

# Make a node df for clade labeling
node.cladelab.df <- data.frame(matrix(nrow=length(unique(tree.bins.info$Subset.Name, ncol=2))))
node.cladelab.df$"Name" <- unique(tree.bins.info$Subset.Name)

for (i in 1:nrow(node.cladelab.df)){
    curr.species <- node.cladelab.df[i, "Name"]
    node.cladelab.df[i, "Name"] <- gsub(curr.species, pattern="s__", replacement="")
    node <- castor::get_mrca_of_set(tree, 
                                    tree.bins.info[tree.bins.info$PlotName==curr.species,
                                                  "tip"])
    node.cladelab.df[i, "node"] <- node
}

# removing these to make it a little easier to read when plotted (no overlapping names)
node.cladelab.df <- node.cladelab.df %>% filter(Name %in% c("Prevotella_colorans",
                                                            "Prevotella_saccharolytica",
                                                            "Prevotella_pleuritidis",
                                                            "Prevotella_micans",
                                                            "Prevotella_bergensis",
                                                            "Prevotella_loescheii",
                                                            "Prevotella_oulorum",
                                                            "Prevotella_marshii")==F)


```

# Plot phylogeny {.tabset}

## Vertical layout
```{r}
p <- ggtree(tree) + 
    theme_tree()


p  %<+% tree.bins.info + 
    geom_point(aes(shape=Found, color=Found
                   ),
               size=5, na.rm=T) +
    geom_text(aes(label=MAG.num)) +
    geom_cladelab(node=node.cladelab.df$node,
                  label=node.cladelab.df$Name, geom="text",
                  offset=0.05, vjust=0, color="black") +
    scale_colour_discrete(na.translate = F) +
    guides(shape = FALSE)
ggsave("../figures/phylogeny_prevotella_MAGS_w_reference.pdf", width=12, height=12)
    
```



# Read in coverage data
Coverage threshold 25% used here: https://www.researchgate.net/publication/316735131_Tracking_microbial_colonization_in_fecal_microbiota_transplantation_experiments_via_genome-resolved_metagenomics/figures?lo=1 

```{r}
coverage.files <- list.files("../prevotella_mags_bowtie", full.names=T)
coverage.files.short <- list.files("../prevotella_mags_bowtie") %>% 
    sapply(function(x) str_split(x,"_")[[1]][1])

insert_fname_to_colnames <- function(df, prefix){
    temp.df <- df
    colnames(temp.df) <- paste0(prefix, "_", colnames(df))
    colnames(temp.df)[1] <- colnames(df)[1] # keep the chr name the same
    return(temp.df)
}

split_bin <- function(rname){
    rname %>% 
        sapply(function(x) str_split(x, "_")[[1]][1])
}

coverage.df <- fread(coverage.files[1])


coverage.df$bin_name <- split_bin(coverage.df$`#rname`)

# Group the coverage info by genome bin
coverage.df <- coverage.df %>% 
    group_by(bin_name) %>%
    summarise(percentcoveredbases=sum((coverage*endpos))/sum(endpos),
              perbasedepth=sum(meandepth*endpos)/sum(endpos))
# rename
coverage.df <- insert_fname_to_colnames(coverage.df, coverage.files.short[1])


for(i in 2:length(coverage.files)){
    file <- coverage.files[i]
    new <- fread(file)
    
    new$bin_name <- split_bin(new$`#rname`)

    # Group the coverage info by genome bin
    new <- new %>% 
        group_by(bin_name) %>%
        summarise(percentcoveredbases=sum((coverage*endpos))/sum(endpos),
                  perbasedepth=sum(meandepth*endpos)/sum(endpos))
    
    new <- insert_fname_to_colnames(new, coverage.files.short[i])
    
    coverage.df <- merge(coverage.df, new, by="bin_name")
}

coverage.df$tip <- coverage.df$bin_name

coverage.df.melted <- coverage.df %>% 
    pivot_longer(cols=c(-tip,-bin_name),
                 names_sep="_",
                 names_to=c("sample","metric"))

covered.bases.columns <- grep("percentcoveredbases", colnames(coverage.df))
coverage.df.covered <- coverage.df[,covered.bases.columns]

coverage.df.matrix <- as.matrix(coverage.df.covered)
colnames(coverage.df.matrix) <- colnames(coverage.df.matrix) %>%
    sapply(function(x) str_split(x, "_")[[1]][1])
rownames(coverage.df.matrix) <- coverage.df$tip

low.risk.samples <- colnames(coverage.df.matrix) %>% 
    sapply(function(x) metadata[x, "Risk_Status"]=="Low")
high.risk.samples <- colnames(coverage.df.matrix) %>% 
    sapply(function(x) metadata[x, "Risk_Status"]=="High")
positive.samples <- colnames(coverage.df.matrix) %>% 
    sapply(function(x) metadata[x, "Risk_Status"]=="Positive")
spacer.cols <- cbind(rep(NA, nrow(coverage.df.matrix)),
                       rep(NA, nrow(coverage.df.matrix)))
colnames(spacer.cols) <- c("Spacer 1", "Spacer 2")

coverage.df.matrix.sorted <- cbind(
      coverage.df.matrix[,low.risk.samples],
      #spacer.cols["Spacer 1"],
      coverage.df.matrix[,high.risk.samples],
      #spacer.cols["Spacer 2"],
      coverage.df.matrix[,positive.samples])
```

# Draw phylogeny with heatmap
```{r}

group <- metadata[colnames(coverage.df.matrix), "Risk_Status"]
group.colors <- group
group.colors[group=="Low"] <- "blue"
group.colors[group=="High"] <- "yellow"
group.colors[group=="Positive"] <- "red"
group.colors[is.na(group)] <- "black"

group <- factor(group, levels=c("Low", "High", "Positive"))

bin.names <- rownames(coverage.df.matrix) %>% 
    sapply(function(x) ifelse(is.na(tree.bins.info[tree.bins.info$tip==x, "PlotName"]),
                              yes=tree.bins.info[tree.bins.info$tip==x, "MAG.num"], 
                              no=paste(tree.bins.info[tree.bins.info$tip==x, "MAG.num"], 
                                       gsub(x=tree.bins.info[tree.bins.info$tip==x, "PlotName"],
                                            pattern="s__", replacement=""),
                                       sep=": "))
    )

heatmap(coverage.df.matrix, ColSideColors=group.colors,
        scale="none", col=pals::coolwarm(),
        labRow=bin.names)

p <- ggtree(tree) + 
    theme_tree()
p <- p %<+% bins.info + 
    geom_label(aes(label = PlotName), nudge_x=0.2) + 
    geom_point(aes(shape=found, color=Known),
               size=5)
gheatmap(p,
         coverage.df.matrix,
         offset=0, width=0.5, font.size=3,
         colnames_angle=-45, hjust=0) +
    scale_fill_continuous()


subset.tree <- drop.tip(tree, 
              tip=tree$tip.label[(tree$tip.label %in% rownames(coverage.df.matrix))==F])
p <- ggtree(subset.tree) + 
    theme_tree()
p <- p %<+% tree.bins.info + 
    geom_label(aes(label = PlotName), nudge_x=0.2) + 
    geom_point(aes(shape=Found, color=Found),

                              size=5) +
    geom_text(aes(label=MAG.num)) +
    scale_colour_discrete(na.translate=F) +
    guides(shape=FALSE, color=FALSE)
    
gheatmap(p,
         coverage.df.matrix.sorted,
         offset=0.05, width=0.5, font.size=3,
         colnames_angle=-45, hjust=0, 
         custom_column_labels=c(rep("Low", sum(low.risk.samples)),
                                #"Spacer 1",
                                rep("High", sum(high.risk.samples)),
                                #"Spacer 2",
                                rep("Positive", sum(positive.samples)))
         ) +
    scale_fill_gradientn(colours = c("blue", "grey", "red"), space = "rgb", 
                         limits = c(0,100))


```
