---
title: "Analysis"
author: "John Sterrett"
date: "2023-1-6"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Note: This  requires microshades, which can be installed using
# remotes::install_github("KarstensLab/microshades")
# and speedyseq, which can be installed here
# remotes::install_github("mikemc/speedyseq")


pacman::p_load(tidyverse,
               data.table,
               ggplot2,
               phyloseq,
               patchwork,
               microshades,
               cowplot)


kable.mod.cols <- c("Estimate", "Std. Error", "deg. freedom", "t value", "p")
```

# Load data
```{r, message=F}
setwd("~/Research-Projects/Lozupone/HIV-MetaG/analysis/")
source("metaphlan_to_phyloseq_helper.R")
bugslist <- fread("../hiv.t32.concat.humann/all_bugs_list.tsv") %>% as.data.frame()
rownames(bugslist) <- bugslist$`#clade_name`
bugslist <- bugslist %>% 
    dplyr::select(-c(`#clade_name`,
               NCBI_tax_id,
               additional_species))

# just get the lowest tax level. We can collapse again later
bugslist <- filter_mphlan_by_taxonomy_level(bugslist, level="max")

# replace NA with 0
bugslist[is.na(bugslist)] <- 0

# check that our data are relative abundance adding up to 100 still
hundreds <- rep(100, ncol(bugslist))
names(hundreds) <- colnames(bugslist)
if (isFALSE(all.equal(target=hundreds, 
                      current=colSums(bugslist, na.rm=T), 
                      tolerance=0.001))){
    print("Data are NOT relative abundances summing to 100! Please check what's going on.")
} else {
    print("Samples sum to 100 (or close enough).")
}

# create tax table
bugs <- rownames(bugslist)
taxonomy.table <- names_to_tax_table(bugs)
rownames(taxonomy.table) <- bugs

# read metadata
metadata <- fread("../metadata.txt") %>% as.data.frame()
rownames(metadata) <- metadata$PID

redcap.DM <- readxl::read_xlsx("../High_Prev_Metadata_redcap.xlsx",sheet=1) %>% as.data.frame()
redcap.LG <- readxl::read_xlsx("../High_Prev_Metadata_redcap.xlsx",sheet=2) %>% as.data.frame()

sex.behav.q <- c("In the past 6 months, what is the frequency at which you have had receptive anal intercourse?",
                 "What gender do you identify with?",
                 "What is your sexual orientation?"
                 )

metadata <- merge(
    metadata,
    rbind(redcap.DM[,c("Record ID", sex.behav.q)],
          redcap.LG[,c("Record ID", sex.behav.q)]),
    by.x="PID",
    by.y="Record ID")
rownames(metadata) <- metadata$PID

metadata$RAI_freq <- metadata$`In the past 6 months, what is the frequency at which you have had receptive anal intercourse?`
metadata$`In the past 6 months, what is the frequency at which you have had receptive anal intercourse?` <- NULL
metadata$RAI <- metadata$RAI_freq!="Never"

metadata$Gender <- metadata$`What gender do you identify with?`
metadata$Orientation <- metadata$`What is your sexual orientation?`
metadata$MSM <- metadata$Gender=="Men" & (metadata$Orientation=="Gay" | metadata$Orientation=="Bisexual")

# address missing data with follow-up from Mallory
metadata[metadata$PID=="DM001", "MSM"] <- TRUE
metadata[metadata$PID=="DM002", "MSM"] <- TRUE
metadata[metadata$PID=="DM007", "MSM"] <- TRUE

metadata[metadata$PID=="DM093", "MSM"] <- FALSE

metadata$MSM.Label <- ifelse(metadata$MSM,
                             yes="MSM",
                             no="MSW")

# create phyloseq object
pseq <- phyloseq(otu_table=otu_table(bugslist, taxa_are_rows=T),
                 tax_table=tax_table(taxonomy.table),
                 sample_data=sample_data(metadata)
                 )

taxonomy.table <- as.data.frame(taxonomy.table)
```

# Unclassified Reads
```{r}
sum(bugs %>% sapply(FUN=function(x) grepl(x, "Unassigned")))
unique(taxonomy.table$Phylum)

eu_un <- bugs[grepl(x=bugs, pattern="Eukaryota_unclassified")]
rowSums(bugslist[eu_un,])/ncol(bugslist)

```

# Taxa barplot {.tabset}
## Custom Microshades legend functions
```{r}
individual_legend <- function (mdf,
                               cdf,
                               group_name,
                               col_name_group = "Top_Phylum",
                               col_name_subgroup = "Top_Genus",
                               x = "Sample",
                               y = "Abundance",
                               legend_key_size = 0.4,
                               legend_text_size = 10)
{
  select_mdf <- mdf %>% filter(!!sym(col_name_group) == group_name)
  select_cdf <- cdf %>% filter(!!sym(col_name_group) == group_name)

  select_plot <- ggplot(select_mdf,
    aes_string(x = x, y = y, fill = col_name_subgroup, text = col_name_subgroup)) +
    geom_col( position="fill") +
    scale_fill_manual(name = group_name,
                      values = select_cdf$hex,
                      breaks = select_cdf[[col_name_subgroup]]) +
    theme(legend.justification = "left") +
    theme(legend.title = element_text(face = "bold")) +
    theme(legend.key.size = unit(legend_key_size, "lines"), text=element_text(size=legend_text_size)) +
    theme(legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent'))

  legend <- get_legend(select_plot)
}

custom_legend <- function (mdf, cdf, group_level = "Phylum", subgroup_level = "Genus", x = "Sample",
                           y = "Abundance", legend_key_size = 0.4, legend_text_size = 10)
{
  if (is.null(mdf[[group_level]])) {
    stop("mdf 'group_level' does not exist")
  }

  if (is.null(mdf[[subgroup_level]])) {
    stop("mdf 'subgroup_level' does not exist")
  }

  if (is.null(cdf$hex)) {
    stop("cdf 'hex' does not exist")
  }

  col_name_group <- paste0("Top_", group_level)
  col_name_subgroup <- paste0("Top_", subgroup_level)

  group_level_names <- unique(cdf[[col_name_group]])

  for (i in 1:length(group_level_names))
  {
    if( i == 1)
    {
      complete_legend <-individual_legend (mdf, cdf, group_level_names[i], col_name_group, col_name_subgroup, legend_key_size = legend_key_size, legend_text_size = legend_text_size)
    }
    else
    {
      new_legend <-individual_legend (mdf, cdf, group_level_names[i], col_name_group, col_name_subgroup, legend_key_size = legend_key_size, legend_text_size =legend_text_size)

      complete_height <- i -1
      new_height <- 1

      complete_legend <-plot_grid(complete_legend, new_legend, ncol = 1, rel_heights = c(complete_height,new_height))
    }
  }
  complete_legend
}

```
## All taxa { .tabset}
### By MSM
```{r, warning=T, fig.height=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Genus")
# sort the phylum names
phylum_table <- tax_glom(pseq, taxrank="Phylum", ) %>% otu_table()
phyla.otunames <- rownames(phylum_table)

phylums <- taxonomy.table[phyla.otunames,"Phylum"]

sorted_phylums <- phylums[order(rowSums(phylum_table), decreasing=T)]
# create the colors object
color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], 
                                  cvd = TRUE)
# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf



# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10) 

# plot
plot <- plot_microshades(mdf_GP, cdf_GP)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7),
        panel.background = element_rect(fill='transparent'), #transparent panel bg
        plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
        panel.grid.major = element_blank(), #remove major gridlines
        panel.grid.minor = element_blank() #remove minor gridlines
        ) + #transparent legend panel
  facet_grid(~MSM.Label, scales="free_x", space="free_x",
             )

multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
ggsave("../figures/taxa-bar-phyum-genus-MSM-microshades.pdf",plot=multi,height=10,width=10, bg="transparent")

```


### By HIV
```{r, warning=T, fig.height=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Genus")
# sort the phylum names
phylum_table <- tax_glom(pseq, taxrank="Phylum", ) %>% otu_table()
phyla.otunames <- rownames(phylum_table)

phylums <- taxonomy.table[phyla.otunames,"Phylum"]

sorted_phylums <- phylums[order(rowSums(phylum_table), decreasing=T)]
# create the colors object
color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], 
                                  cvd = TRUE)
# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10)

# plot
plot <- plot_microshades(mdf_GP, cdf_GP)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~HIV_Status, scales="free_x", space="free_x"
             #labeller=labeller(Osteosarcoma=ost.names)
             )

multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
ggsave("../figures/taxa-bar-phyum-genus-microshades.pdf",plot=multi,height=10,width=10)

```

### By Risk Status
```{r, warning=T, fig.height=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Genus")
# sort the phylum names
phylum_table <- tax_glom(pseq, taxrank="Phylum", ) %>% otu_table()
phyla.otunames <- rownames(phylum_table)

phylums <- taxonomy.table[phyla.otunames,"Phylum"]

sorted_phylums <- phylums[order(rowSums(phylum_table), decreasing=T)]
# create the colors object
color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], 
                                  cvd = TRUE)
# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10)

# plot
plot <- plot_microshades(mdf_GP, cdf_GP)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~(factor(Risk_Status, levels=c("Low", "High", "Positive"))), 
               scales = "free_x", space="free_x"
             )

multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
ggsave("../figures/taxa-bar-Risk-phyum-genus-microshades.pdf",plot=multi,height=10,width=10)

```


## All taxa - Phylum (species microshades) with expanded prevotella {.tabset}
### by MSM
```{r,  warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Species")

# create the colors object
# was going to force prevotella in, but it's already the top
color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], group_level="Phylum", subgroup_level="Species",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Phylum", "Species", "Bacteroidetes",
                              existing_palette="micro_cvd_purple", new_palette = "micro_purple",
                              n_add=2)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Phylum",
                          subgroup_level="Species")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  theme(axis.text.x = element_text(size= 7),
        panel.background = element_rect(fill='transparent'), #transparent panel bg
        plot.background = element_rect(fill='transparent', color=NA), #transparent plot bg
        panel.grid.major = element_blank(), #remove major gridlines
        panel.grid.minor = element_blank() #remove minor gridlines
        ) +
  facet_grid(~MSM.Label, scales = "free_x", space="free_x"
             )

multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
ggsave("../figures/taxa-bar-phyum-species-expanded-prev-MSM-microshades.pdf",plot=multi,height=10,width=10, bg="transparent")
```


### by HIV
```{r,  warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Species")

# create the colors object
# was going to force prevotella in, but it's already the top
color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], group_level="Phylum", subgroup_level="Species",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Phylum", "Species", "Bacteroidetes",
                              existing_palette="micro_cvd_purple", new_palette = "micro_purple",
                              n_add=2)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Phylum",
                          subgroup_level="Species")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~HIV_Status, scales = "free_x", space="free_x"
             #labeller=labeller(Osteosarcoma=ost.names)
             )

multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
#ggsave("../figures/taxa-bar-phyum-genus-microshades.pdf",plot=multi,height=6,width=10)

```

### by Risk status
```{r,  warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Species")

# create the colors object
# was going to force prevotella in, but it's already the top
color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], group_level="Phylum", subgroup_level="Species",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Phylum", "Species", "Bacteroidetes",
                              existing_palette="micro_cvd_purple", new_palette = "micro_purple",
                              n_add=2)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Phylum",
                          subgroup_level="Species")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~(factor(Risk_Status, levels=c("Low", "High", "Positive"))), 
               scales = "free_x", space="free_x"
             )

multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
#ggsave("../figures/taxa-bar-phyum-genus-microshades.pdf",plot=multi,height=6,width=10)
```

## All taxa - Genus (species microshades) with expanded prevotella {.tabset}
### By HIV
```{r, warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Species")

# create the colors object
# was going to force prevotella in, but it's already the top

# sort the genus names
genus_table <- tax_glom(pseq, taxrank="Genus", ) %>% otu_table()
genus.otunames <- rownames(genus_table)

genera <- taxonomy.table[genus.otunames,"Genus"]

sorted_genera <- genera[order(rowSums(genus_table), decreasing=T)]

color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_genera[5:1], group_level="Genus", subgroup_level="Species",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Genus", "Species", "Prevotella",
                              existing_palette="micro_cvd_purple", new_palette = "micro_purple",
                              n_add=4)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf

#make other other black
#cdf_GP[cdf_GP$hex=="#F5F5F5", "hex"] <- "#000000"
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Genus",
                          subgroup_level="Species")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP,)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~HIV_Status, scales = "free_x", space="free_x"
             #labeller=labeller(Osteosarcoma=ost.names)
             )

  # facet_grid(Phylum~HIV_Status, scales = "free", space="free_x"
  #            #labeller=labeller(Osteosarcoma=ost.names)
  #            )


multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
#ggsave("../figures/taxa-bar-phyum-genus-microshades.pdf",plot=multi,height=6,width=10)

```

### By Risk Status
```{r, warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(pseq, subgroup_level="Species")

# create the colors object
# was going to force prevotella in, but it's already the top

# sort the genus names
genus_table <- tax_glom(pseq, taxrank="Genus", ) %>% otu_table()
genus.otunames <- rownames(genus_table)

genera <- taxonomy.table[genus.otunames,"Genus"]

sorted_genera <- genera[order(rowSums(genus_table), decreasing=T)]

color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_genera[5:1], group_level="Genus", subgroup_level="Species",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Genus", "Species", "Prevotella",
                              existing_palette="micro_cvd_purple", new_palette = "micro_purple",
                              n_add=4)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf

#make other other black
#cdf_GP[cdf_GP$hex=="#F5F5F5", "hex"] <- "#000000"
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Genus",
                          subgroup_level="Species")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP,)
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~(factor(Risk_Status, levels=c("Low", "High", "Positive"))), 
               scales = "free_x", space="free_x"
             )


multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
#ggsave("../figures/taxa-bar-phyum-genus-microshades.pdf",plot=multi,height=6,width=10)
```


# Prevotella specifically
```{r}
prev.only <- subset_taxa(pseq, Genus=="Prevotella") %>% 
    tax_glom(taxrank="Species") 
prev.only.df <- prev.only %>% otu_table() %>% as.data.frame()
prev.only.tax <- prev.only %>% tax_table() %>% as.data.frame()

# grab just the species level
splitted <- lapply(rownames(prev.only.df),
      FUN=function(x) strsplit(x, split="\\|")[[1]][7])
# clean name and reassign rowname
splitted <- gsub("[a-z]__", "", splitted)
rownames(prev.only.df) <- splitted

# calculate total relative abundance of prevotella in each sample
total.prev <- colSums(prev.only.df)

# calculate total relative abundance of prevotella copri in each sample
copri.names <- splitted[grepl("copri", splitted)]
total.copri <- colSums(prev.only.df[copri.names,])

# calculate total non-copri prevotella
total.noncopri <- total.prev - total.copri

prev.stats <- data.frame("Prevotella"=total.prev, 
                         "Prevotella_copri"=total.copri, 
                         "Prevotella_noncopri"=total.noncopri)

prev.stats <- merge(prev.stats, metadata[c("HIV_Status", "%Prevotella", "Risk_Status")], by=0)
prev.stats$expected_Prevotella <- prev.stats$`%Prevotella`
prev.stats$HIV <- prev.stats$HIV_Status
prev.stats$`%Prevotella` <- NULL
prev.stats$HIV_Status <- NULL

rownames(prev.stats) <- prev.stats$Row.names
prev.stats$Row.names <- NULL

metadata[rownames(prev.stats[prev.stats$expected_Prevotella < 0.05,]), ]
metadata[rownames(prev.stats[prev.stats$Prevotella < 5,]), ]

```


## Prevotella ~ expected (from 16S) {.tabset}
### LM and plot
```{r}

ggplot(data=prev.stats,
       mapping=aes(y=Prevotella, x=expected_Prevotella*100)) +
    geom_point() + 
    geom_text(label=rownames(prev.stats)) +
    geom_smooth(method="lm") +
    ylab("% Prevotella (metagenomes)") +
    xlab("% Prevotella (16S)") +
    theme_bw()  

mod.prev.expected <- lm(Prevotella ~ expected_Prevotella, data=prev.stats)
summary(mod.prev.expected)

resids <- mod.prev.expected$residuals
lowest_resids <- resids[order(resids)][1:5]
print(lowest_resids)

```


### Taxa bar plot of lower than expected Prevotella {.tabset}
#### By HIV
```{r, warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(prune_samples(samples=names(lowest_resids), x=pseq), 
                     subgroup_level="Genus")

# create the colors object
# was going to force prevotella in, but it's already the top


color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], group_level="Phylum", subgroup_level="Genus",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Phylum", "Genus", "Firmicutes",
                              existing_palette="micro_cvd_green", new_palette = "micro_green",
                              n_add=4)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf

#make other other black
#cdf_GP[cdf_GP$hex=="#F5F5F5", "hex"] <- "#000000"
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Phylum",
                          subgroup_level="Genus")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP,) + theme_bw()
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~HIV_Status, scales = "free_x", space="free_x"
             #labeller=labeller(Osteosarcoma=ost.names)
             )


multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi

```

#### By Risk Status
```{r, warning=T, fig.height=9, fig.width=8}
# prep the microshades colors
mdf_prep <- prep_mdf(prune_samples(samples=names(lowest_resids), x=pseq), 
                     subgroup_level="Genus")

# create the colors object
# was going to force prevotella in, but it's already the top


color_objs_GP <- create_color_dfs(mdf_prep, selected_groups = sorted_phylums[5:1], group_level="Phylum", subgroup_level="Genus",
                                  cvd = TRUE,)
color_objs_GP <- extend_group(color_objs_GP$mdf, color_objs_GP$cdf,
                              "Phylum", "Genus", "Firmicutes",
                              existing_palette="micro_cvd_green", new_palette = "micro_green",
                              n_add=4)

# Extract
mdf_GP <- color_objs_GP$mdf
cdf_GP <- color_objs_GP$cdf

#make other other black
#cdf_GP[cdf_GP$hex=="#F5F5F5", "hex"] <- "#000000"
# create a custom legend
GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                          legend_key_size=unit(0.4, "cm"),
                          legend_text_size=10,
                          group_level="Phylum",
                          subgroup_level="Genus")

# plot
plot <- plot_microshades(mdf_GP, cdf_GP,) + theme_bw()
plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
  theme(legend.position = "none")  +
  theme(axis.text.x = element_text(size= 7)) +
  facet_grid(~(factor(Risk_Status, levels=c("Low", "High", "Positive"))), 
             scales = "free_x", space="free_x"
             )


multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
multi
```

### Firmicutes investigation
```{r}
firm.only <- subset_taxa(pseq, Phylum=="Firmicutes") %>% 
    tax_glom(taxrank="Order") 

firm.only.weird.samps <- firm.only %>% prune_samples(samples=names(lowest_resids))

firm.only.weird.samps.otu <- firm.only.weird.samps %>% otu_table()


splitted <- lapply(rownames(firm.only.weird.samps.otu),
      FUN=function(x) strsplit(x, split="\\|")[[1]][1:4])
# clean name and reassign rowname
splitted <- lapply(splitted, FUN=function(x) gsub("[a-z]__", "", x))
joined <- lapply(splitted, FUN=function(x) paste(x, collapse=" "))
rownames(firm.only.weird.samps.otu) <- joined

firm.only.weird.samps.otu[order(rowSums(firm.only.weird.samps.otu), decreasing=T),] %>% 
    as.data.frame() %>%
    head(n=15) %>% 
    knitr::kable()

```

## Relative abundance boxplots {.tabset}
### Prevotella {.tabset}
#### by HIV
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella, x=HIV)) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw()

```

#### by Risk Status
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella, 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw() +
    xlab("Risk Status") +
    ylab("% Prevotella")
```


### Prevotella copri {.tabset}
#### Relative abundance (HIV)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_copri, x=HIV)) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw()
```

#### Relative abundance (Risk Status)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_copri, 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw() +
    xlab("Risk Status") +
    ylab("% Prevotella copri")
```


#### As percent of all Prevotella (HIV)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_copri/(Prevotella+0.001), x=HIV)) +
    geom_boxplot(outlier.shape=NA) + 
    geom_jitter(width=0.1) +
    theme_bw()
```

#### As percent of all Prevotella (Risk Status)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_copri/(Prevotella+0.001), 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw() +
    xlab("Risk Status")
```

### Prevotella non-copri {.tabset}
#### Relative abundance (HIV)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_noncopri, x=HIV)) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw()
```

#### Relative abundance (Risk Status)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_noncopri, 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw() +
    xlab("Risk Status") +
    ylab("% other Prevotella")
```

#### As percent of all Prevotella (HIV)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_noncopri/(Prevotella+0.001), x=HIV)) +
    geom_boxplot(outlier.shape=NA) + 
    geom_jitter(width=0.1) +
    theme_bw()
```

#### As percent of all Prevotella (Risk Status)
```{r}
ggplot(data=prev.stats,
       mapping=aes(y=Prevotella_noncopri/(Prevotella+0.001), 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot() + 
    geom_jitter(width=0.1) +
    theme_bw() +
    xlab("Risk Status")
```


# Diversity of Prevotella {.tabset}
## Calculating diversity
```{r}
observed.prev.feats <- colSums(prev.only.df > 0)
observed.pcopri.feats <- colSums(prev.only.df[copri.names,] > 0 )

p.div <- data.frame(observed.pcopri.feats, observed.prev.feats)
p.div <- merge(p.div, metadata[c("HIV_Status", "%Prevotella", "Risk_Status")], by=0)
p.div$expected_Prevotella <- p.div$`%Prevotella`
p.div$HIV <- p.div$HIV_Status
p.div$`%Prevotella` <- NULL
p.div$HIV_Status <- NULL

rownames(p.div) <- p.div$Row.names
p.div$Row.names <- NULL

```

## Boxplot of Observed Prevotella features ~ Risk status
```{r}
ggplot(data=p.div,
       mapping=aes(y=observed.prev.feats, 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot(outlier.shape=NA) + 
    geom_jitter(width=0.1, height=0, size=3) +
    theme_bw() +
    xlab("Risk Status") +
    ylab("Observed Prevotella features")


```


## Boxplot of Observed P copri features ~ Risk status
```{r}
ggplot(data=p.div,
       mapping=aes(y=observed.pcopri.feats, 
                   x=factor(Risk_Status, levels=c("Low", "High", "Positive")))) +
    geom_boxplot(outlier.shape=NA) + 
    geom_jitter(width=0.1, height=0, size=3) +
    theme_bw() +
    xlab("Risk Status") +
    ylab("Observed Prevotella copri clades")

```
